vpr/src/route/lookahead_profiler.cpp

```diff
@@ -0,0 +1,173 @@
+//
+// Created by shrevena on 08/06/24.
+//
+
+#include <sstream>
+
+#include "globals.h"
+#include "lookahead_profiler.h"
+#include "re_cluster_util.h"
+#include "router_lookahead_map_utils.h"
+#include "vpr_utils.h"
+#include "vtr_error.h"
+#include "vtr_log.h"
+
+void LookaheadProfiler::record(int iteration,
+                               int target_net_pin_index,
+                               const t_conn_cost_params& cost_params,
+                               const RouterLookahead& router_lookahead,
+                               const ParentNetId& net_id,
+                               const Netlist<>& net_list,
+                               const std::vector<RRNodeId>& branch_inodes) {
+    auto& device_ctx = g_vpr_ctx.device();
+    const auto& rr_graph = device_ctx.rr_graph;
+    auto& route_ctx = g_vpr_ctx.routing();
+
+    // If csv file hasn't been opened, open it and write out column headers
+    if (is_empty) {
```

Comment:
The only issue is that the profiler object is initialized in ``RoutingContext``, so if opening the file is done in the constructor, it will happen even if the option to use it isn't on

Reply:
I'm not sure how to do this without creating a constructor for RoutingContext and VprContext as well; would this be okay?

Reply:
You're right. RoutingContext has a global instance, meaning that LookaheadProfiler's constructor is called before parsing command line options.

I think we shouldn avoid adding anything to the global context unless it is really necessary. Instead, we could ceate a LookaheadProfiler instance at the start of the routing stage and pass it through a function call hierarchy down to its call site.

Reply:
If you decide to keep LookaheadProfiler instance global, you should free the memory allocated for its member variables after it is no longer needed. This is the global instance doesn't go out of scope